File: ece1724-project/graphx-test.scala
import org.apache.spark.graphx._
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession

object GraphXTest {
  def main(args: Array[String]): Unit = {
    // Create SparkSession
    val spark = SparkSession.builder
      .appName("GraphX Example")
      .master("local[*]") // Adjust master for cluster setups
      .getOrCreate()

    val sc = spark.sparkContext

    // Load vertices
    val vertices: RDD[(VertexId, String)] = sc.parallelize(Seq(
      (1L, "Alice"),
      (2L, "Bob"),
      (3L, "Charlie"),
      (4L, "David"),
      (5L, "Eve"),
      (6L, "Frank"),
      (7L, "Grace"),
      (8L, "Hank")
    ))

    // Load edges
    val edges: RDD[Edge[Int]] = sc.parallelize(Seq(
      Edge(1L, 2L, 1), // Alice -> Bob
      Edge(2L, 3L, 1), // Bob -> Charlie
      Edge(3L, 1L, 1), // Charlie -> Alice
      Edge(3L, 4L, 1), // Charlie -> David
      Edge(4L, 5L, 1), // David -> Eve
      Edge(5L, 6L, 1), // Eve -> Frank
      Edge(6L, 4L, 1), // Frank -> David
      Edge(7L, 8L, 1), // Grace -> Hank
      Edge(8L, 7L, 1)  // Hank -> Grace
    ))

    // Create the graph
    val graph = Graph(vertices, edges)

    // Print vertices
    println("Vertices:")
    graph.vertices.collect.foreach(println)

    // Print edges
    println("Edges:")
    graph.edges.collect.foreach(println)

    // Run PageRank
    println("Running PageRank...")
    val ranks = graph.pageRank(0.0001).vertices
    println("PageRank Results:")
    ranks.collect.foreach(println)

    // Stop SparkSession
    spark.stop()
  }
}

File: GraphX-data/graphx-test.py
from pyspark.sql import SparkSession
from graphframes import GraphFrame

# Initialize SparkSession with the GraphFrames JAR
spark = SparkSession.builder \
    .appName("GraphFramesExample") \
    .config("spark.jars", "/usr/local/spark/jars/graphframes-0.8.3-spark3.4-s_2.12.jar") \
    .getOrCreate()

# Create vertices and edges DataFrames
vertices = spark.createDataFrame([
    ("1", "Alice"),
    ("2", "Bob"),
    ("3", "Charlie"),
    ("4", "David"),
    ("5", "Eve"),
    ("6", "Frank"),
    ("7", "Grace"),
    ("8", "Hank")
], ["id", "name"])

edges = spark.createDataFrame([
    ("1", "2"),  # Alice -> Bob
    ("2", "3"),  # Bob -> Charlie
    ("3", "1"),  # Charlie -> Alice
    ("3", "4"),  # Charlie -> David
    ("4", "5"),  # David -> Eve
    ("5", "6"),  # Eve -> Frank
    ("6", "4"),  # Frank -> David
    ("7", "8"),  # Grace -> Hank
    ("8", "7")   # Hank -> Grace
], ["src", "dst"])

# Create a GraphFrame
graph = GraphFrame(vertices, edges)

# Run PageRank
results = graph.pageRank(resetProbability=0.15, maxIter=10)
results.vertices.show()

print("Running PageRank...")
results = graph.pageRank(resetProbability=0.15, maxIter=10)

print("PageRank Results (Vertices with Ranks):")
results.vertices.select("id", "name", "pagerank").show()